contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,204 @@
+// for each read
+providerId, keyId, base64Encoded := // slice provider and key from value
+
+// ensure this provider is the one handling providerId
+aead := // lookup an aead instance for keyId or error
+bytes := base64.Decode(base64Encoded)
+nonce, authenticatedData, cipherText := // slice from bytes
+out, err := aead.Open(nil, nonce, cipherText, authenticatedData)
+```
+
+## Key Generation & Rotation
+
+To account for easy key rotation, an additional layer of abstraction is introduced where keys used for resource encryption is encrypted by another set of keys (Key Encryption Keys aka KEK).
+The key used for encrypting kubernetes resources (Data Encryption Keys, aka DEK) are generated by the apiserver and stored in etcd once encrypted with one of the KEKs.
+KEKs should never be stored in etcd and in most cases be on the same volume as the apiserver. This provides consistency for the apiserver but still protects against compromises to the etcd volume.
+
+In order to take an API driven approach for key rotation, new API objects will be defined:
+* Key Encryption Key (KEK) - key used to unlock the Data Encryption Key. Stored on API server nodes.
+* Data Encryption Key (DEK) - long-lived secret encrypted with a KEK. Stored in etcd encrypted. Unencrypted in-memory in API servers.
+* KEK Slot - to support KEK rotation there will be an ordered list of KEKs stored in the KEK DB. The current active KEK slot number, is stored in etcd for consistency.
+* KEK DB - a file with N KEKs in a JSON list. KEK[0], by definition, is null.
+
+```go
+type DataEncryptionKey struct {
+    ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Value string    // Encrypted
+}
+```
+
+```go
+type KeyEncryptionKeySlot struct {
+    ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Slot int
+}
+```
+
+```go
+type KeyEncryptionKeyDatabase struct {
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    Keys []string
+}
+```
+
+To enable encryption a user will issue a PUT to an endpoint such as `/rotate`. If this is the first time this API has ever been called the API server
+will generate a key (unecrypted DEK), encrypt it with the KEK in slot 1, and encrypt all secrets with the DEK. If this is the second + N time the API
+has been called the API server will encrypt the DEK with the KEK in slot N+1 and do a compare-and-swap on the DEK stored in etcd.
+If this API is called and a KEK doesn't exist in the KEK DB at slot N+1 a 400 is returned. `kubectl` will have support for calling the
+rotate API endpoint eventually. To account for failure scenarios during key rotation, the old and new DEK will be stored in etcd during the rotation.
+
+## Master Configuration
+In order to enable encryption, a user must first create a KEK DB file and tell the API server to use it with --key-encryption-key-db-path=/path/to/kekdb/file. The file will be a simple YAML file that lists all of the keys:
```

Comment:
I'm not sure on this one. It might make sense for the resources to encrypt to be outside the encryption provider. Would putting the resources inside the encryption provider mean the encryption provider would have to be configured for all resources so that it could decide if it should or shouldn't encrypt?